### PR TITLE
feat: use email validation link 

### PR DESCRIPTION
**Pull request by danielgent**
~This isn't actually tested yet, I figure I'll do that next and there'll be another PR with fixes~

~*I could also* start writing tests for what I've done so far if we're happy with this route~

I've ran this locally and tested it! this does actually set `email_verified` to true!

#### Next steps
* should I do the account linking on this same route? I think that makes sense :thinking: 
* write tests for the work I've done so far? I think we're pretty set on doing this... 
* have the validate-email continue the login flow? this is definitely more complex and I'd rather avoid this if possible :thinking:

### CODE REVIEW

File: src/controllers/email.ts [hunk 1]
```diff
@@ -24,59 +24,6 @@ function getLocale(language: string) {
 
 const engine = new Liquid();
 
-// exportasync function sendEmailValidation(
```
Comment (danielgent) [inline src/controllers/email.ts:27]: lol. I recreated this method below on the previous PR. Proof that commenting out methods for later is totally pointless :joy:

File: src/adapters/interfaces/Users.ts [hunk 2]
```diff
@@ -12,10 +12,6 @@ export interface UserDataAdapter {
   create(tenantId: string, user: User): Promise<User>;
   remove(tenantId: string, id: string): Promise<boolean>;
   list(tenantId: string, params: ListParams): Promise<ListUsersResponse>;
-  update(
-    tenantId: string,
-    id: string,
-    user: Partial<BaseUser>,
-  ): Promise<boolean>;
+  update(tenantId: string, id: string, user: Partial<User>): Promise<boolean>;
```
Comment (danielgent) [inline src/adapters/interfaces/Users.ts:15]: Seems like we want to have all the fields in `User` here *BUT* we're not exposing this to the outside world

File: src/controllers/email.ts [hunk 3]
```diff
@@ -293,7 +240,7 @@ export async function sendValidateEmailAddress(
   const emailValidationBody = await engine.render(sendEmailValidationTemplate, {
     // TODO - what does auth0 do here? Haven't actually seen it in action... Would be good to copy their route
     // need to actually create this route!
-    emailValidationUrl: `${env.ISSUER}u/validate-email-address?state=${state}&code=${code}`,
+    emailValidationUrl: `${env.ISSUER}u/validate-email?state=${state}&code=${code}`,
```
Comment (danielgent) [inline src/controllers/email.ts:243]: matches with the route we already have on login

File: src/routes/tsoa/login.ts [hunk 4]
```diff
@@ -306,25 +305,29 @@ export class LoginController extends Controller {
       throw new HTTPException(400, { message: "Client not found" });
     }
 
-    try {
-      // another duplicate here - DRY rule of three!
-      const otps = await env.data.OTP.list(client.tenant_id, email);
```
Comment (danielgent) [inline src/routes/tsoa/login.ts:311]: matches the forgot password data store. as discussed, we have multiple for no specific reason :smile:

File: src/routes/tsoa/dbconnections.ts [hunk 5]
```diff
@@ -12,26 +12,14 @@ import { RequestWithContext } from "../../types/RequestWithContext";
...
-  AuthorizationResponseType,
-  AuthParams,
-  CodeChallengeMethod,
-} from "../../types";
+import { AuthParams } from "../../types";
 import generateOTP from "../../utils/otp";
 
-// duplicated from /passwordless route - EXTRACT THIS OUT!
-const CODE_EXPIRATION_TIME = 30 * 60 * 1000;
+const CODE_EXPIRATION_TIME = 24 * 60 * 60 * 1000;
```
Comment (danielgent) [inline src/routes/tsoa/dbconnections.ts:22]: as discussed, a full day to validate this *BUT* we'll have to think of a different solution if we want the more complex rule of _continue login flow if code is less than 30 minutes old else just validate email_ :smile:

File: src/controllers/email.ts [hunk 6]
```diff
@@ -271,7 +218,7 @@ export async function sendValidateEmailAddress(
   state: string,
 ) {
   const response = await env.AUTH_TEMPLATES.get(
-    "templates/email/validate-email-address.liquid",
+    "templates/email/verify-email.liquid",
```
Comment (danielgent) [inline src/controllers/email.ts:221]: ooops! as mentioned above. this should work now

File: src/email-templates/verify-email.mjml
Comment (danielgent) [inline src/email-templates/verify-email.mjml:1]: this was previously a copy paste of the code email, just to add to the confusion :+1:

File: src/adapters/interfaces/Codes.ts [hunk 7]
```diff
@@ -2,5 +2,5 @@ import { Code } from "../../types/Code";
 
 export interface CodesAdapter {
   create: (tenant_id: string, authCode: Code) => Promise<void>;
-  list: (tenant_id: string, email: string) => Promise<Code[]>;
+  list: (tenant_id: string, user_id: string) => Promise<Code[]>;
```
Comment (danielgent) [inline src/adapters/interfaces/Codes.ts:5]: this was a fun bug! the param on the adapter was called `email`, but the actual query was by `user_id` :smile:

File: src/controllers/email.ts [hunk 8]
```diff
@@ -293,7 +240,7 @@ export async function sendValidateEmailAddress(
   const emailValidationBody = await engine.render(sendEmailValidationTemplate, {
     // TODO - what does auth0 do here? Haven't actually seen it in action... Would be good to copy their route
     // need to actually create this route!
```
Comment (danielgent) [inline src/controllers/email.ts:242]:
```suggestion
    // we have not checked the route name that auth0 uses
```

File: src/routes/tsoa/login.ts [hunk 9]
```diff
@@ -306,25 +305,29 @@ export class LoginController extends Controller {
...
-      }
+    const user = await getUserByEmailAndProvider({
+      userAdapter: env.data.users,
+      tenant_id: client.tenant_id,
+      email,
+      provider: "auth2",
+    });
+    if (!user) {
+      throw new HTTPException(500, { message: "No user found" });
+    }
```
Comment (danielgent) [inline src/routes/tsoa/login.ts:316]: This route is tested and working now. I had to move around the order of calls a bit

File: src/routes/tsoa/login.ts [hunk 10]
```diff
@@ -306,25 +305,29 @@ export class LoginController extends Controller {
...
 
-      return handleLogin(env, this, user, session);
-    } catch (err: any) {
-      return renderEmailValidation(env, this, session, err.message);
+    if (!foundCode) {
+      throw new HTTPException(400, { message: "Code not found or expired" });
     }
+
+    await env.data.users.update(client.tenant_id, user.id, {
+      email_verified: true,
```
Comment (danielgent) [inline src/routes/tsoa/login.ts:326]: tested and works!

File: src/routes/tsoa/login.ts [hunk 11]
```diff
@@ -306,25 +305,29 @@ export class LoginController extends Controller {
...
+    });
+
+    // what should we actually do here?
+    return "email validated";
```
Comment (danielgent) [inline src/routes/tsoa/login.ts:330]: *next* should I do the account linking here? :thinking:

**Review by markusahlstrand — APPROVED** [review]
LGTM!

**Pull request merged**